### PR TITLE
DrivAerML: 5L/256d+T_max=30 — deeper architecture

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

4L/256d at T_max=30 achieved the current best of 12.96% — a massive improvement over 3L/192d (33.65%). If depth is the critical lever, does 5L/256d continue the trend? We observed 3L→4L was a dramatic step function improvement. 5L may yield further gains or reveal a diminishing-returns inflection point. 6L was found to diverge.

## Instructions

Run the same command as the DrivAerML baseline but add `--model-layers 5`:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-5L256d
```

**Key change:** `--model-layers 5` (was 4). All other hyperparameters identical to current best.

**Watch for:** OOM errors — if the 5L model is too large, report the error and we will adjust. If it trains, report val_primary/surface_rel_l2_pct at every epoch.

## Baseline

- **Dataset:** DrivAerML
- **Primary metric:** `val_primary/surface_rel_l2_pct` (lower is better)
- **Current best:** **12.96%** (val) / 14.41% (test) — PR #2550
- **Best W&B run:** 8s5i8y06 (violet — 4L/256d, T_max=30, 43 epochs, AdamW lr=5e-4)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**CRITICAL flags (required to avoid OOM):**
- `--batch-size 1`
- `--drivaerml-train-surface-points 50000`
- `--drivaerml-eval-surface-points 50000`
- `--max-train-batches 394`
- `--max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### Summary

| Model | val_primary/surface_rel_l2_pct | test_primary/surface_rel_l2_pct | Epochs | W&B Run |
|---|---|---|---|---|
| 4L/256d baseline (#2550) | **12.96%** | 14.41% | 43 | 8s5i8y06 |
| **5L/256d (this PR)** | **13.24%** | **14.16%** | 45 | 9fwg8o17 |

### Epoch-by-Epoch Progress (selected)

| Epoch | val_primary (%) | loss | lr |
|---|---|---|---|
| 1 | 61.64 | 0.828 | 2.16e-5 |
| 5 | 38.05 | 0.311 | 3.75e-4 |
| 10 | 29.57 | 0.196 | 1.25e-4 |
| 15 | 25.63 | 0.140 | 0 |
| 20 | 23.50 | 0.107 | 1.25e-4 |
| 25 | 21.17 | 0.082 | 3.75e-4 |
| 30 | 18.55 | 0.068 | 5.0e-4 |
| 35 | 15.71 | 0.052 | 3.75e-4 |
| 40 | 13.83 | 0.042 | 1.25e-4 |
| 43 | 14.04 | 0.043 | 8.27e-5 |
| 45 | **13.24** | 0.041 | 0 |

### Command Used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent shoya --wandb-group drivaerml-5L256d --wandb-name "shoya/5L256d-tmax30"
```

### Peak Memory / Timing
- GPU: NVIDIA RTX PRO 6000 Blackwell (98GB)
- ~6.8 GB VRAM peak
- 30m41s wall time, 45 epochs completed

### What happened

**5L/256d essentially ties the 4L/256d baseline** — no clear improvement from adding a 5th layer:
- **Val: 13.24% vs 12.96%** — 5L is 0.28 pp worse
- **Test: 14.16% vs 14.41%** — 5L is 0.25 pp better
- The difference is within noise. At the same epoch count (43), 5L had val=14.04% vs 4L's 12.96%.

**The 5L model was still converging at cutoff.** Loss dropped from 0.828 to 0.041 over 45 epochs, with each cosine trough delivering a new val best. The best val (13.24%) came at the final epoch (45, lr=0 trough). With more epochs, 5L could potentially beat 4L — but at the same compute budget, 4L is the better choice.

**Convergence pattern:** Clear cosine-restart oscillation. Best val metrics consistently appear at LR troughs. The gap between trough-best and restart-worst narrows over training (epoch 8: 34→39, epoch 40: 14→17), indicating the model is becoming more robust to LR perturbations.

**No OOM issues.** The 5L/256d model runs comfortably within the 6.8GB VRAM budget at batch_size=1 with 50K surface points.

### Assessment

**Verdict: Diminishing returns from depth.** The 3L→4L jump was dramatic (-61.5% relative improvement). The 4L→5L jump is negligible (~0 improvement at matched epochs). This suggests we've hit the depth sweet spot at 4 layers for this architecture/dataset/epoch-budget. Note: 6L was previously found to diverge, so 5L may be the maximum stable depth.

### Suggested follow-ups

1. **Keep 4L/256d as the DrivAerML architecture** — 5L doesn't justify the extra compute.
2. **Try width scaling instead** — since depth has diminishing returns, try 4L/384d or 4L/512d (if they fit in memory).
3. **Longer training for 4L** — the 4L baseline was still converging at epoch 43. More epochs is likely the highest-ROI path to beating 12.96%.
4. **Tune T_max** — the cosine restart frequency (T_max=30) may not be optimal. Larger T_max means fewer restarts and more sustained learning per cycle.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2605 because ADVISOR: 5L/256d gives 13.24% — consistent with violet's 13.62% finding. 4L→5L adds no benefit (matched-epoch val: 14.04% vs 4L's 12.96%). Depth sweet spot confirmed at 4 layers. DEAD END..
